### PR TITLE
Add WAMP Test Vectors / Conformance Test Suite

### LINKED ISSUE #556: WAMP Test Vectors / Conformance Test Suite
# WAMP Test Vectors / Conformance Test Suite

## Summary

This proposal aims to create a comprehensive, machine-readable conformance test suite for WAMP implementations. The test suite would consist of **test vectors** - canonical examples of WAMP messages and interaction sequences that implementations can use to validate correctness and interoperability.

## Motivation

Currently, WAMP implementations validate correctness primarily through:
1. Manual testing during development
2. Ad-hoc integration tests
3. Examples embedded in the specification (human-readable only)

While the WAMP specification includes excellent examples throughout the documentation, these are:
- **Human-readable only** - cannot be directly consumed by automated tests
- **Scattered** - distributed across different sections of the spec
- **Incomplete** - not every feature has comprehensive examples
- **Implementation-specific** - each implementation creates its own test data

**This is brilliant and would be transformative for WAMP implementations!** A machine-readable conformance test suite would fundamentally change how WAMP implementations are developed, tested, and validated.

### 1. Cross-Implementation Compatibility
- If Autobahn|JS, Autobahn|Python, Autobahn|Java, and other implementations all pass the same test vectors, they are **provably interoperable**
- Eliminates "works on my implementation" issues
- Provides confidence for mixing implementations in production

### 2. Regression Prevention
- Implementations can run test suite in CI/CD
- Changes that break conformance are caught immediately
- Version updates maintain compatibility

### 3. New Implementer Guidance
- Clear, executable examples of "what a valid CHALLENGE message looks like"
- Reduces ambiguity in specification
- Lowers barrier to entry for new implementations

### 4. Specification Precision
- Writing machine-readable examples forces precision
- Uncovers edge cases and ambiguities
- Makes implicit assumptions explicit

### 5. Certification & Compliance
- Potential for "WAMP Certified" badge for implementations passing full test suite
- Users can evaluate implementation maturity objectively
- Encourages ecosystem quality

## Comprehensive Feature Coverage

The test suite should cover all WAMP features systematically:

### Core Protocol
- All WAMP message types (HELLO, WELCOME, PUBLISH, EVENT, CALL, RESULT, ERROR, etc.)
- Basic Publish & Subscribe
- Basic Remote Procedure Calls
- Session lifecycle (authentication, goodbye)

### Advanced Features
- **Publish & Subscribe**:
  - Publisher Identification
  - Publisher Exclusion / Inclusion
  - Subscriber Black/Whitelisting
  - Pattern-based Subscriptions
  - Subscription Revocation
  - Event History
  - Acknowledged Event Delivery

- **Remote Procedure Calls**:
  - Progressive Call Results
  - Caller Identification
  - Pattern-based Registrations
  - Shared Registrations
  - Registration Revocation
  - Call Timeout
  - Call Cancellation

- **Authentication**:
  - WAMP-CRA
  - WAMP-Ticket
  - WAMP-Cryptosign
  - Cookie authentication
  - TLS certificate authentication

- **Transports**:
  - Message Batching
  - WebSocket compression
  - Payload Passthru Mode

- **Payload Formats**:
  - JSON
  - MessagePack
  - CBOR
  - Multiple arguments formats (args, kwargs)
  - Binary payloads

### Edge Cases
- Empty arguments
- Null values
- Large payloads
- Nested structures
- Binary data preservation
- Unicode handling

## Proposed Structure

```
wamp-proto/
└── testsuite/
    ├── README.md                    # Documentation for test suite format and usage
    ├── SCHEMA.json                  # JSON Schema for test vector format
    │
    ├── singlemessage/               # Individual WAMP message serialization tests
    │   ├── basic/                   # Basic Profile messages
    │   │   ├── hello.json
    │   │   ├── welcome.json
    │   │   ├── abort.json
    │   │   ├── goodbye.json
    │   │   ├── error.json
    │   │   ├── publish.json
    │   │   ├── published.json
    │   │   ├── subscribe.json
    │   │   ├── subscribed.json
    │   │   ├── unsubscribe.json
    │   │   ├── unsubscribed.json
    │   │   ├── event.json
    │   │   ├── call.json
    │   │   ├── result.json
    │   │   ├── register.json
    │   │   ├── registered.json
    │   │   ├── unregister.json
    │   │   ├── unregistered.json
    │   │   ├── invocation.json
    │   │   └── yield.json
    │   │
    │   └── advanced/                # Advanced Profile features
    │       ├── publisher_identification.json
    │       ├── publisher_exclusion.json
    │       ├── publisher_inclusion.json
    │       ├── subscriber_blacklisting.json
    │       ├── subscriber_whitelisting.json
    │       ├── pattern_based_subscription.json
    │       ├── subscription_revocation.json
    │       ├── event_history.json
    │       ├── acknowledged_event_delivery.json
    │       ├── progressive_call_results.json
    │       ├── caller_identification.json
    │       ├── pattern_based_registration.json
    │       ├── shared_registration.json
    │       ├── registration_revocation.json
    │       ├── call_timeout.json
    │       ├── call_cancellation.json
    │       ├── payload_passthru.json
    │       └── ...
    │
    ├── singlesession/               # Single client session interaction sequences
    │   ├── basic/
    │   │   ├── session_establishment.json      # HELLO → WELCOME
    │   │   ├── subscribe_and_event.json        # SUBSCRIBE → SUBSCRIBED → EVENT
    │   │   ├── call_and_result.json            # CALL → RESULT
    │   │   ├── call_error.json                 # CALL → ERROR
    │   │   └── session_close.json              # GOODBYE → GOODBYE
    │   │
    │   └── advanced/
    │       ├── progressive_results.json        # CALL → RESULT (progress) → RESULT (final)
    │       ├── authenticated_session.json      # HELLO → CHALLENGE → AUTHENTICATE → WELCOME
    │       └── ...
    │
    ├── multisession/                # Multiple client sessions (requires router)
    │   ├── basic/
    │   │   ├── pubsub_basic.json              # Publisher publishes, subscriber receives
    │   │   ├── rpc_basic.json                 # Caller invokes, callee receives and yields
    │   │   └── ...
    │   │
    │   └── advanced/
    │       ├── publisher_exclusion_enabled.json   # Publisher does NOT receive own EVENT
    │       ├── publisher_exclusion_disabled.json  # Publisher DOES receive own EVENT
    │       ├── subscriber_whitelist.json          # Only whitelisted subscribers receive EVENT
    │       ├── shared_registration.json           # Multiple callees share registration
    │       └── ...
    │
    └── multirouter/                 # Multiple routers (federation, clustering)
        ├── router_to_router_link.json
        ├── cross_realm_pubsub.json
        └── ...
```

## Test Vector Format

Each test vector file should be a JSON document with the following structure:

```json
{
  "description": "EVENT message with publisher identification",
  "wamp_message_type": "EVENT",
  "wamp_message_code": 36,
  "feature": "advanced.pubsub.publisher_identification",
  "spec_reference": "https://wamp-proto.org/wamp_latest_ietf.html#name-publisher-identification",

  "serializers": {
    "json": {
      "bytes": "[36, 5512315355, 4429313566, {\"publisher\": 1234}, [], {\"color\": \"orange\"}]",
      "bytes_hex": "5b33362c20353531323331353335352c20343432393331333536362c207b227075626c6973686572223a20313233347d2c205b5d2c207b22636f6c6f72223a20226f72616e6765227d5d"
    },
    "msgpack": {
      "bytes_hex": "96241a00000000"
    }
  },

  "expected_attributes": {
    "message_type": 36,
    "subscription_id": 5512315355,
    "publication_id": 4429313566,
    "details": {
      "publisher": 1234
    },
    "args": [],
    "kwargs": {
      "color": "orange"
    }
  },

  "validation": {
    "python": {
      "import": "from autobahn.wamp.message import Event",
      "assertions": [
        "msg.subscription == 5512315355",
        "msg.publication == 4429313566",
        "msg.publisher == 1234",
        "msg.kwargs['color'] == 'orange'"
      ]
    },
    "javascript": {
      "assertions": [
        "msg.subscription === 5512315355",
        "msg.publication === 4429313566",
        "msg.publisher === 1234",
        "msg.kwargs.color === 'orange'"
      ]
    }
  },

  "notes": [
    "Publisher identification must be explicitly enabled in subscription options",
    "The publisher field in Details is the WAMP session ID of the publisher"
  ]
}
```

### For Multi-Message Sequences

```json
{
  "description": "Publisher exclusion disabled - publisher receives own EVENT",
  "feature": "advanced.pubsub.publisher_exclusion",
  "spec_reference": "https://wamp-proto.org/wamp_latest_ietf.html#name-publisher-exclusion",

  "sessions": [
    {
      "session_id": "session_1",
      "roles": ["publisher", "subscriber"]
    }
  ],

  "sequence": [
    {
      "step": 1,
      "description": "Client subscribes to topic",
      "from": "session_1",
      "to": "router",
      "message": {
        "type": "SUBSCRIBE",
        "request_id": 123,
        "topic": "com.myapp.topic1"
      }
    },
    {
      "step": 2,
      "description": "Router confirms subscription",
      "from": "router",
      "to": "session_1",
      "message": {
        "type": "SUBSCRIBED",
        "request_id": 123,
        "subscription_id": 456
      }
    },
    {
      "step": 3,
      "description": "Client publishes with exclude_me=false",
      "from": "session_1",
      "to": "router",
      "message": {
        "type": "PUBLISH",
        "request_id": 789,
        "options": {
          "exclude_me": false
        },
        "topic": "com.myapp.topic1",
        "args": ["Hello, world!"]
      }
    },
    {
      "step": 4,
      "description": "Publisher receives EVENT because exclude_me=false",
      "from": "router",
      "to": "session_1",
      "message": {
        "type": "EVENT",
        "subscription_id": 456,
        "publication_id": 999,
        "details": {},
        "args": ["Hello, world!"]
      },
      "assertion": "Publisher MUST receive EVENT when exclude_me=false"
    }
  ],

  "expected_outcome": {
    "session_1_receives_event": true,
    "reason": "exclude_me option was explicitly set to false"
  }
}
```

### Phase 1: Foundation
1. Create `testsuite/` directory structure
2. Define JSON schema for test vector format
3. Write comprehensive README with usage guidelines
4. Create initial test vectors for core message types

### Phase 2: Basic Coverage
5. Complete all Basic Profile message types
6. Add validation for message structure
7. Provide multiple serialization formats (JSON, MessagePack)

### Phase 3: Advanced Features
8. Add Advanced Profile test vectors
9. Include authentication mechanisms
10. Cover transport-specific features

### Phase 4: Multi-Session Sequences
11. Define format for interaction sequences
12. Create test vectors for common scenarios
13. Document expected router behavior

### Phase 5: Ecosystem Integration
14. Update specification to reference test vectors
15. Integrate into existing WAMP implementation test suites
16. Create test runner / validator tool (optional)

### For Client Library Developers

```python
import json
from autobahn.wamp import serializer

# Load test vector
with open('testsuite/singlemessage/basic/event.json') as f:
    test_vector = json.load(f)

# Test deserialization
json_serializer = serializer.JsonSerializer()
bytes_data = test_vector['serializers']['json']['bytes'].encode('utf-8')
msg = json_serializer.deserialize(bytes_data)

# Validate attributes
expected = test_vector['expected_attributes']
assert msg.subscription == expected['subscription_id']
assert msg.publication == expected['publication_id']
assert msg.kwargs == expected['kwargs']

# Test serialization (roundtrip)
serialized = json_serializer.serialize(msg)
msg2 = json_serializer.deserialize(serialized)
assert msg == msg2
```

### For Router Developers

```python
# Load multi-session test vector
with open('testsuite/multisession/advanced/publisher_exclusion_disabled.json') as f:
    test_vector = json.load(f)

# Simulate the message sequence
for step in test_vector['sequence']:
    if step['from'] == 'session_1' and step['to'] == 'router':
        # Process incoming message from client
        router.process_message(step['message'])
    elif step['from'] == 'router':
        # Verify router sends expected message
        sent_msg = router.get_next_outgoing_message(step['to'])
        assert sent_msg == step['message']

# Verify expected outcome
assert test_vector['expected_outcome']['session_1_receives_event'] == True
```

## Benefits to the WAMP Ecosystem

This conformance test suite would:

1. **Improve Quality**: Implementations can validate correctness automatically
2. **Increase Interoperability**: Cross-implementation compatibility is provable
3. **Lower Barrier to Entry**: New implementations have clear examples to follow
4. **Clarify Specification**: Machine-readable examples reduce ambiguity
5. **Enable Certification**: Potential for "WAMP Compliant" certification
6. **Prevent Regressions**: CI/CD integration catches breaking changes
7. **Document Best Practices**: Canonical examples serve as reference implementations

## Open Questions

1. **Format**: Is JSON the right format for test vectors, or should we consider YAML, TOML, or protocol buffers?
2. **Scope**: Should we start with Basic Profile only, or include Advanced Profile from the beginning?
3. **Serializers**: Which serialization formats should be included in test vectors? (JSON, MessagePack, CBOR, etc.)
4. **Validation**: Should we provide a reference test runner, or leave that to implementations?
5. **Versioning**: How do we handle test vectors for different WAMP specification versions?
6. **Maintenance**: Who maintains the test suite? Should it be part of the spec release process?

## Next Steps

If this proposal is accepted, the next steps would be:

1. Finalize test vector format and JSON schema
2. Create initial directory structure
3. Write comprehensive documentation (README.md)
4. Generate initial test vectors from existing specification examples
5. Create sample implementation usage in Autobahn|Python
6. Gather feedback from other WAMP implementers
7. Iterate and expand coverage

## References

- [WAMP Specification](https://wamp-proto.org/wamp_latest_ietf.html)
- Similar test vector approaches:
  - [Wycheproof - Google's Crypto Test Vectors](https://github.com/google/wycheproof)
  - [JSON-LD Test Suite](https://json-ld.org/test-suite/)
  - [WebAuthn Test Vectors](https://github.com/w3c/webauthn)

---

**This proposal aims to elevate WAMP implementations to a new level of conformance, stability, and reliability.**

Feedback and suggestions are welcome!

### PR DESCRIPTION
**Pull request by oberstet**
fixes https://github.com/wamp-proto/wamp-proto/issues/556

### CODE REVIEW

**Comment by oberstet** [comment]
here is an example test with autobahn-python (see https://github.com/crossbario/autobahn-python/pull/1765):

```
oberstet@amd-ryzen5:~/work/wamp/autobahn-python$ just test-serdes cpy311
...
Successfully installed autobahn-25.11.1
==> No venv name specified. Auto-detecting from system Python...
==> Defaulting to venv: 'cpy312'
==> Running WAMP message serdes conformance tests in cpy312...
==> Test vectors loaded from: wamp-proto/testsuite/
===================================================== test session starts =====================================================
platform linux -- Python 3.12.11, pytest-9.0.1, pluggy-1.6.0
rootdir: /home/oberstet/work/wamp/autobahn-python
configfile: pytest.ini
plugins: asyncio-1.3.0, aiohttp-1.1.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 28 items                                                                                                            

examples/serdes/tests/test_publish.py .s....s....s...s...sss......                                                      [100%]

=================================================== short test summary info ===================================================
SKIPPED [1] examples/serdes/tests/test_publish.py:66: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_publish.py:110: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_publish.py:148: Flatbuffers with transparent payload not supported (issue #1766)
SKIPPED [4] examples/serdes/tests/test_publish.py:205: Flatbuffers with transparent payload not supported (issue #1766)
================================================ 21 passed, 7 skipped in 0.09s ================================================
(cpy312) oberstet@amd-ryzen5:~/work/wamp/autobahn-python$ 
```

**Comment by oberstet** [comment]
added a design analysis / review document, going into WAMP End-to-End Encryption and Router-to-Router Messaging:

https://github.com/wamp-proto/wamp-proto/blob/9db5436b22c200801729fdeb0e94774ab07a4255/wep/WIP_E2EE_AND_RLINKS.md

**Comment by oberstet** [comment]
test integration - here for autobahn-python - works now: https://github.com/crossbario/autobahn-python/pull/1765#issuecomment-3535161557

here is local testing:

```
...
Successfully installed autobahn-25.11.1
==> No venv name specified. Auto-detecting from system Python...
==> Defaulting to venv: 'cpy312'
==> Running WAMP message serdes conformance tests in cpy312...
==> Test vectors loaded from: wamp-proto/testsuite/
===================================================== test session starts =====================================================
platform linux -- Python 3.12.11, pytest-9.0.1, pluggy-1.6.0
rootdir: /home/oberstet/work/wamp/autobahn-python
configfile: pytest.ini
plugins: asyncio-1.3.0, aiohttp-1.1.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 172 items                                                                                                           

examples/serdes/tests/test_publish.py .s....s....s...s...sss........................................................... [ 47%]
..................                                                                                                      [ 57%]
examples/serdes/tests/test_event.py .s....s....s...s...sss...................................................           [100%]

=================================================== short test summary info ===================================================
SKIPPED [1] examples/serdes/tests/test_publish.py:67: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_publish.py:113: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_publish.py:151: Flatbuffers with transparent payload not supported (issue #1766)
SKIPPED [1] examples/serdes/tests/test_publish.py:220: Serializer flatbuffers not in test vector
SKIPPED [3] examples/serdes/tests/test_publish.py:216: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_event.py:67: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_event.py:113: Serializer flatbuffers not in test vector
SKIPPED [1] examples/serdes/tests/test_event.py:151: Flatbuffers with transparent payload not supported (issue #1766)
SKIPPED [1] examples/serdes/tests/test_event.py:221: Serializer flatbuffers not in test vector
SKIPPED [3] examples/serdes/tests/test_event.py:217: Serializer flatbuffers not in test vector
=============================================== 158 passed, 14 skipped in 0.20s ===============================================
oberstet@amd-ryzen5:~/work/wamp/autobahn-python$ 
```

**Comment by oberstet** [comment]
alright, now all WAMP message types - BP and AP - are covered:

- Phase 1 (Pub/Sub): 8 message types (7 basic + 1 advanced)
- Phase 2 (RPC): 10 message types (8 basic + 2 advanced)
- Phase 3 (Shared): 1 message type
- Phase 4 (Session Lifecycle): 6 message types
- Total: 25 WAMP message types

**Comment by om26er** [comment]
This is a super interesting effort and something we should participate in as well with our implementations @Mahad-10 @muzzammilshahid @rubyonrails3

File: testsuite/README.md [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+    },
+    "args": [],
+    "kwargs": {
+      "color": "orange"
+    }
+  },
+
+  "validation": {
+    "python": {
+      "import": "from autobahn.wamp.message import Event",
```
Comment (KSDaemon) [inline testsuite/README.md]: I think this is a bit incorrect: why is validation for Python based on the autobahn implementation?
I think this should be an implementation-free thing. 

* There may be more than 1 implementation in one language.
* There may be a bug in the autobahn.wamp.message too.

Either this part should also be in a machine-readable format without any implementation-specific code.

Or it should be done like in `Multiple WAMP Implementation Support` below: 

instead of:

```json
"python": {
      "import": "from autobahn.wamp.message import Event",
      "assertions": [
        "msg.subscription == 5512315355",
        "msg.publication == 4429313566",
        "msg.publisher == 1234",
        "msg.kwargs['color'] == 'orange'"
      ]
    },
```

it should be:

```json
"autobahn.python": {
      "import": "from autobahn.wamp.message import Event",
      "assertions": [
        "msg.subscription == 5512315355",
        "msg.publication == 4429313566",
        "msg.publisher == 1234",
        "msg.kwargs['color'] == 'orange'"
      ]
    },
```

Instead of:
```json
    "javascript": {
      "assertions": [
        "msg.subscription === 5512315355",
        "msg.publication === 4429313566",
        "msg.publisher === 1234",
        "msg.kwargs.color === 'orange'"
      ]
    }
```

it should be:

```json
    "autobahn.js": {
      "assertions": [
        "msg.subscription === 5512315355",
        "msg.publication === 4429313566",
        "msg.publisher === 1234",
        "msg.kwargs.color === 'orange'"
      ]
    }
    "wampy": {
      "assertions": [
        "msg.options.subscription === 5512315355",
        "msg.options.publication === 4429313566",
        "msg.options.publisher === 1234",
        "msg.color === 'orange'"
      ]
    }
```

**Review by KSDaemon — COMMENTED** [review]
This looks great and important! We've been talking about it for a long time! And @oberstet finally got a time to write it down! cool!

File: testsuite/README.md (hunk 1 above)
Comment (oberstet) [reply testsuite/README.md]: absolutely! good catch! I agree, this should also be using the "multi WAMP impl" approach - which supports multiple different WAMP impls even for a single host language - https://github.com/wamp-proto/wamp-proto/blob/f64e466b47de7d8c271b8964cbe76e5076c95ea8/testsuite/README.md#multiple-wamp-implementation-support. gimme bit time, I'll fix this!

Comment (oberstet) [reply testsuite/README.md]: fixed:

  ✅ Fixed "Single Message Test Vector" Example

  Before (WRONG - language-based indexing):
```
  "validation": {
    "python": {           // ❌ Generic language name
      "import": "...",
      "assertions": [...]
    },
    "javascript": {       // ❌ Generic language name
      "assertions": [...]
    }
  }
```
  // ❌ Missing "construction" section entirely

  After (CORRECT - implementation-based indexing):
```
  "validation": {
    "autobahn-python": [  // ✅ Specific WAMP implementation
      "from autobahn.wamp.message import Event\n..."
    ],
    "autobahn-js": [      // ✅ Specific WAMP implementation
      "// Test framework provides 'msg'...\n"
    ],
    "wampy": [            // ✅ Another implementation (shows multi-impl support)
      "// Wampy.js uses different structure...\n"
    ]
  },

  "construction": {       // ✅ Added missing section
    "autobahn-python": "from autobahn.wamp.message import Event\n..."
  }
```

  Why This Matters:

  1. Multiple implementations per language: Python has autobahn-python, wampy-python, etc.
  2. Different APIs: Each implementation has different message class structures
  3. Avoids implementation bias: Not assuming autobahn is "the" Python implementation
  4. Matches actual files: publish.json, event.json, call.json, etc. already use this pattern

  Verified:

  - ✅ All 5 existing test vectors with validation/construction already use correct structure
  - ✅ "Multiple WAMP Implementation Support" section was already correct
  - ✅ Addresses KSDaemon's (wampy.js) feedback from PR #557

and:

  ✅ Verified Structure Alignment:

  Top Level:

```
  {
    "description": "MESSAGE_TYPE message...",  // Message type description
    "wamp_message_type": "MESSAGE_TYPE",
    "wamp_message_code": NN,
    "feature": "...",
    "spec_reference": "...",
    "samples": [ ... ],                         // ✅ Array wrapper
    "notes": [ ... ]
  }
```

  Inside samples[i]:

```
  {
    "description": "...",                       // ✅ Sample-specific description

    "serializers": {                            // ✅ Indexed by serializer
      "json": [{ ... }],
      "msgpack": [{ ... }],
      "cbor": [{ ... }],
      "ubjson": [{ ... }]
    },

    "expected_attributes": {                    // ✅ Dict (non-indexed)
      "message_type": NN,
      "request_id": 123,
      ...
    },

    "validation": {                             // ✅ Indexed by WAMP implementation
      "autobahn-python": [ "..." ],
      "autobahn-js": [ "..." ],
      "wampy": [ "..." ]
    },

    "construction": {                           // ✅ Indexed by WAMP implementation
      "autobahn-python": "...",
      "autobahn-js": "..."
    }
  }
```

  All requirements met:
  - ✅ samples.description: string attribute
  - ✅ samples.expected_attributes: dict attribute (non-indexed)
  - ✅ samples.serializers: indexed by serializer
  - ✅ samples.validation: indexed by WAMP implementation
  - ✅ samples.construction: indexed by WAMP implementation

Comment (oberstet) [reply testsuite/README.md]: curious: is anyone interested in WAMP-Flatbuffers? and respective coverage in the testsuite on this PR?

I have added that to autobahn-python via https://github.com/crossbario/autobahn-python/pull/1765, that's not yet released on PyPI, but on GitHub https://github.com/crossbario/autobahn-python/discussions/1768

the schemata are included in the Python package, they are here in the source tree: https://github.com/crossbario/autobahn-python/tree/master/autobahn/wamp/flatbuffers

fwiw, here are some related pieces from the autobahn-docs:

image

image

**Comment by oberstet** [comment]
so WAMP-Flexbuffers now has full coverage too! eg see: https://github.com/crossbario/autobahn-python/pull/1773#issuecomment-3568805213

File: .github/ISSUE_TEMPLATE/config.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+blank_issues_enabled: false
+contact_links:
+  - name: 💬 Discussions
+    url: https://github.com/crossbario/autobahn-python/discussions
+    about: Ask questions and discuss ideas with the community
+  - name: 📖 Documentation
+    url: https://crossbar.io/docs/
+    about: Read the official documentation
```
Comment (KSDaemon) [inline .github/ISSUE_TEMPLATE/config.yml:8]: I know - you love crossbar :) But how is it related to being present here?)

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+1. [First step]
+2. [Second step]
+3. [...]
+
+**Minimal Reproducible Example**:
+```python
+# Code that reproduces the issue
+```
+
```
Comment (KSDaemon) [inline .github/ISSUE_TEMPLATE/bug_report.md:29]:
```suggestion

Or link to a gist.

```

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
+---
```
Comment (KSDaemon) [inline .github/ISSUE_TEMPLATE/bug_report.md:1]: I thought about it, templates in general....
While I love them and find them very useful for providing structured information, what I see from my experience, not many issue creators will fill them as we expect... Probably only some parts will be filled, while others will be kept blank. And looking at those issues with just templated text is really annoying. 

So, as a suggestion: Maybe it would be better to wrap each template section into comments (< !--  -- >), because when you click `Create issue` the editor is opened in edit mode with the md content. And those who really want to fill - can make it and uncomment the filled section, while not filled - keep as is, but in rendered mode - we will see only the filled one. In this way, we can save the time of issue creator, and the time of those who will be looking at it later (without useless unfilled template parts).

WDYT?

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+**Related Issues**:
+- #XXX
+- #YYY
+
+## Checklist
+
+- [ ] I have searched existing issues to avoid duplicates
+- [ ] I have provided a minimal reproducible example
+- [ ] I have included version information
+- [ ] I have included error messages/logs
```
Comment (KSDaemon) [inline .github/ISSUE_TEMPLATE/bug_report.md:62]: We can also set up CI to check this list and if none is marked - gently ask the creator to revisit this PR (and maybe add some labels for us like `not filled` or smth like that)

**Review by KSDaemon — COMMENTED** [review]
Maybe I missed smth, but a few things are not clear to me [yet]: 
You have exact `id`s encoded in test vectors and examples. While in reality it might be difficult to set/receive exactly the same ids, as it's an internal property of the wamp implementations that are in most cases exposed as a `read-only` to the client. So it won't be possible to take and compare as is. 

So maybe instead we should use some form of `PLACEHOLDER` for such properties, explaining that yes, it should exist, but it is autogenerated, so please don't compare the exact values here. Because we want to test not only the message serializer but rather the whole execution flow.

Folks, WDYT?

File: .github/ISSUE_TEMPLATE/config.yml (hunk 2 above)
Comment (oberstet) [reply .github/ISSUE_TEMPLATE/config.yml:8]: yes, absolutely! filed as https://github.com/wamp-proto/wamp-cicd/issues/4

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 3 above)
Comment (oberstet) [reply .github/ISSUE_TEMPLATE/bug_report.md:29]: ah yeah, love gists ..

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 4 above)
Comment (oberstet) [reply .github/ISSUE_TEMPLATE/bug_report.md:1]: I agree! I would say: if we cannot automatically check it, such as for example "requiring always N-out-of-M check boxes checked", then wrap template example text (and it is nothing else then) into comments (otherwise it's just annoying)

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 5 above)
Comment (oberstet) [reply .github/ISSUE_TEMPLATE/bug_report.md:62]: that's a great list! I'll add that on the PR!

we can still decide how to auto process it:

i) deny even _submitting_ the issue
ii) fail workflow check job (once we get CI active again)
iii) label issue "needs context" (manual work required)
...

File: .github/ISSUE_TEMPLATE/config.yml (hunk 2 above)
Comment (oberstet) [reply .github/ISSUE_TEMPLATE/config.yml:8]: I should have been more explicit: those links come into this PR for wamp-proto as it instantiates https://github.com/wamp-proto/wamp-cicd as a git submodule, but the github issue templates that come in via that are _not properly templated_ with respect to the _using_ repo - which is wamp-proto in this case (this PR)

**Comment by oberstet** [comment]
> Maybe I missed smth, but a few things are not clear to me [yet]: You have exact `id`s encoded in test vectors and examples. While in reality it might be difficult to set/receive exactly the same ids, as it's an internal property of the wamp implementations that are in most cases exposed as a `read-only` to the client. So it won't be possible to take and compare as is.
> 
> So maybe instead we should use some form of `PLACEHOLDER` for such properties, explaining that yes, it should exist, but it is autogenerated, so please don't compare the exact values here. Because we want to test not only the message serializer but rather the whole execution flow.
> 
> Folks, WDYT?

yes, I agree! here is my take on it:

_currently_ this PR still only covers **singlemessage** samples, a single WAMP, either a valid one, or an invalid one.

in this case, IDs can be checked properly by a test suite run, always, independent of WAMP implementation specifics, as IDs are not in any way interpreted across multiple WAMP peers or sessions

once we come to **multisession** test suite samples, this will be different, a single sample will contain multiple WAMP message, e.g. `SUBSCRIBE` and `SUBSCRIBED` involves a WAMP Subscriber and a (single) WAMP Router, and **certain IDs in this case must match accross messages**, eg in

```
[SUBSCRIBED, SUBSCRIBE.Request|id, Subscription|id]
```

the `SUBSCRIBE.Request|id` _must_ match the one from

```
[SUBSCRIBE, Request|id, Options|dict, Topic|uri]
```

https://wamp-proto.org/wamp_latest_ietf.html#name-subscribe-2
https://wamp-proto.org/wamp_latest_ietf.html#name-subscribed-2

**this is exactly what WAMP protocol semantics define**, and that is the essence of the protocol itself.

-----

as a consequence, since the test suite cannot know IDs instantiated/used by a WAMP implementation, and by a single _run_ of said implementation, the samples in the testsuite will need to use **placeholder IDs**, eg. `ID9|id`

and when the testsuite runs, it needs to:

1. upon processing _the first_ message matching `ID9|id` having say `24827`, _remember*_ that `ID9 == 24827` from now on
2. once processing a subsequent message with (in the sample) `ID9|id`, now check for equality with `24827` as of _*_

----

this strategy can be expanded

* across multiple WAMP sessions too!
* it can even be expanded across multiple WAMP router nodes - for testing router-to-router
* it can be be expanded across regular WAMP sessions, and management sessions a WAMP router might expose for *WAMP meta event & procedures*

of course the proof is in the pudding, so _demonstrating_ this with real WAMP implementations is required, and I'm sure "some detail questions" _will_ come up only then. 

what do you guys think?

File: .github/ISSUE_TEMPLATE/bug_report.md (hunk 5 above)
Comment (KSDaemon) [reply .github/ISSUE_TEMPLATE/bug_report.md:62]: I think we can not prevent submitting such an issue, as it is just the issue description. But do an issue check right after submitting and comment and label it - is definitely what we can do! :)
